### PR TITLE
Types for the plugins

### PR DESCRIPTION
**Pull request by christiancp100**
### What does it do?

Describe the technical changes you did.

### Why is it needed?

Describe the issue you are solving.

### How to test it?

Provide information about the environment and the path to verify the behaviour.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

**Comment by Bassel17** [comment]
is the upgrade ts-zen change supposed to be in this branch ?

**Comment by innerdvations** [comment]
I just refactored everything to match the documentation, so @christiancp100 @Bassel17 please double check the parts you worked on and make sure I didn't break anything.

**Comment by innerdvations** [comment]
Note on the failing lint Server: It is not our exported Config.Server, it's actually caused by `packages/core/strapi/src/services/server/http-server.ts` so it will be fixed in the parent branch

File: packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import type { Context } from 'koa';
+import { Strapi } from '../../../../..';
+
+export type ControllerReturnType = {
```
Comment (innerdvations) [inline packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts:4]: We shouldn't be using our own type here, we should be using the same type used by createCoreController https://github.com/strapi/strapi/blob/361b318f91416ef5984112ac5627eaff7988d750/packages/core/strapi/src/factories.ts#L22-L27

File: packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  [key: string]: (ctx: Context) => Promise<unknown> | unknown;
+};
+
+export type Controller = ({ strapi }: { strapi: Strapi }) => ControllerReturnType;
```
Comment (innerdvations) [inline packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts:8]: We shouldn't be using our own type here, we should be using the same type used by createCoreController https://github.com/strapi/strapi/blob/361b318f91416ef5984112ac5627eaff7988d750/packages/core/strapi/src/factories.ts#L22-L27

Comment (innerdvations) [reply packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts:8]: This way it adds the strapi callback and also ensures that it only includes the extendable methods

Comment (Bassel17) [reply packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts:8]: am not sure I understand we don't have uid here

Comment (Bassel17) [reply packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts:8]: we don't have any predefined controllers or extendables,
plugin creators can define what they want

Comment (innerdvations) [reply packages/core/typings/src/types/core/plugins/config/strapi-server/controllers.ts:8]: Ah, true, sorry. But we're still re-inventing the wheel a bit. I'll take a look in the morning, it feels like there's something we can do.

File: packages/core/typings/src/types/core/plugins/config/strapi-server/routes.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import type { Common } from '../../..';
+// import { Router } from '../../../../core-api';
```
Comment (innerdvations) [inline packages/core/typings/src/types/core/plugins/config/strapi-server/routes.ts:2]: we can remove this

**Review by innerdvations — APPROVED** [review]
LGTM! Side note for anyone looking in, we will be adding factories or some other way to assert these types very soon.

**Comment by mojofawad** [comment]
> is the upgrade ts-zen change supposed to be in this branch ?

Same question. ts-zen was breaking my GitHub Actions

**Pull request merged**